src/FTPServer.jl

```diff
@@ -0,0 +1,178 @@
+)
+
+function __init__()
+    Memento.register(LOGGER)
+    copy!(pyopenssl_crypto, pyimport_conda("OpenSSL.crypto", "OpenSSL"))
+    copy!(pyopenssl_SSL, pyimport_conda("OpenSSL.SSL", "OpenSSL"))
+    copy!(pyftpdlib_servers, pyimport_conda("pyftpdlib.servers", "pyftpdlib", "invenia"))
+
+    DEBUG && pylogging[:basicConfig](level=pylogging[:DEBUG])
+    mkpath(ROOT)
+end
+
+
+mutable struct Server
+    root::AbstractString
+    port::Int
+    username::AbstractString
+    password::AbstractString
+    permissions::AbstractString
+    security::Symbol
+    process::Process
+    io::IO
+end
+
+"""
+    Server(root=$ROOT; username=$USER, password=$PASSWD, permissions=$PERM, security=:none)
+
+A Server stores settings for create an pyftpdlib server.
+
+# Arguments
+- `root::AbstractString`: Directory where you want store credentials and data for the
+  test server.
+
+# Keywords
+- `username`: Default login username
+- `password`: Default login password
+- `permission`: Default user read/write permissions
+- `security`: Security method to use for connecting (options: `:none`, `:implicity`, `:explicity`).
+  Passing in `:none` will use FTP and passing in `:implicity` or `:explicity` will use the appropriate
+  FTPS connection.
+"""
+function Server(
+    root::AbstractString=ROOT; username="", password="", permissions="elradfmwM",
+    security::Symbol=:none,
+)
+    if isempty(username)
+        username = string("user", rand(1:9999))
+    end
+    if isempty(password)
+        password = randstring(40)
+    end
+
+    cmd = `$PYTHON_CMD $SCRIPT $username $password $root --permissions $permissions`
+    if security != :none
+        cmd = `$cmd --tls $security --cert-file $CERT --key-file $KEY --gen-certs TRUE`
+    end
+    io = Pipe()
+
+    # Note: open(::AbstractCmd, ...) won't work here as it doesn't allow us to capture STDERR.
+    process = if VERSION > v"0.7.0-DEV.4445"
+        run(pipeline(cmd, stdout=io, stderr=io), wait=false)
+    else
+        spawn(pipeline(cmd, stdout=io, stderr=io))
+    end
+
+    line = readline(io)
+    m = match(r"starting FTP.* server on .*:(?<port>\d+)", line)
+    if m !== nothing
+        port = parse(Int, m[:port])
+        Server(root, port, username, password, permissions, security, process, io)
+    else
+        kill(process)
+        error(line, String(readavailable(io)))  # Display traceback
+    end
+end
+
+"""
+    serve(f, args...; kwargs...)
+
+Passes `args` and `kwargs` to the `Server` constructor and runs the function `f` by passing
+in the `server` instance. Upon completion the `server` will automatically be shutdown.
+"""
+function serve(f::Function, args...; kwargs...)
+    server = Server(args...; kwargs...)
+
+    try
+        f(server)
+    finally
+        close(server)
+    end
+end
+
+hostname(server::Server) = "localhost"
+port(server::Server) = server.port
+username(server::Server) = server.username
+password(server::Server) = server.password
```

Comment:
A `uri(server::Server) = "ftp://$(username(server)):$(password(server))@$(hostname(server)):$(port(server))"` would be useful.